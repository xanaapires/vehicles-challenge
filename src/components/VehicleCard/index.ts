export { default as VehicleCard } from "./VehicleCard";
