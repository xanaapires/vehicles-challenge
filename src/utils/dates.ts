import moment from "moment";

export const getRelativeTime = (start: Date) => {
  const now = moment();
  const target = moment(start);

  const isPast = target.isBefore(now);
  const duration = moment.duration(Math.abs(target.diff(now)));

  const days = Math.floor(duration.asDays());
  const hours = duration.hours();

  return isPast
    ? `${days} day(s) and ${hours} hour(s) ago`
    : `in ${days} day(s) and ${hours} hour(s)`;
};
