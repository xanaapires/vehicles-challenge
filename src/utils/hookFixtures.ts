import { vehiclesData } from "../assets/vehiclesData";

export const defaultStateMock = {
  vehiclesToShow: vehiclesData,
  filteredVehicles: vehiclesData,
  filters: {
    make: "",
    model: "",
    startBid: "",
    endBid: "",
    favouritesOnly: false,
  },
  applyFilter: () => {},
  availableMakes: [
    {
      label: "All",
      value: "all",
    },
    {
      label: "Audi",
      value: "Audi",
    },
    {
      label: "BMW",
      value: "BMW",
    },
    {
      label: "Citroen",
      value: "Citroen",
    },
    {
      label: "Ford",
      value: "Ford",
    },
    {
      label: "Mercedes-Benz",
      value: "Mercedes-Benz",
    },
    {
      label: "Toyota",
      value: "Toyota",
    },
    {
      label: "Volkswagen",
      value: "Volkswagen",
    },
    {
      label: "Volvo",
      value: "Volvo",
    },
  ],
  availableModels: [],
  bidValues: [],
  getVehicleById: () => {},
  toggleFavourite: () => {},
};
