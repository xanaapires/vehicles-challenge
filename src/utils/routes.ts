export const ROUTES = {
  LISTING: "/",
  VEHICLE_DETAILS: "/vehicle/:vehicleId",
};
