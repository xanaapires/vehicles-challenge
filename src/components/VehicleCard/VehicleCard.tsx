import React, { useCallback } from "react";
import { Vehicle } from "../../types";
import { Box, Button, Divider, IconButton, Typography } from "@mui/material";
import StarBorderOutlinedIcon from "@mui/icons-material/StarBorderOutlined";
import StarOutlinedIcon from "@mui/icons-material/StarOutlined";
import TimerOutlinedIcon from "@mui/icons-material/TimerOutlined";
import { defaultColors, fontWeight } from "../ThemeProvider/themeConfigs";
import emptyPhoto from "../../assets/empty_photo.jpg";
import moment from "moment";
import { getRelativeTime } from "../../utils/dates";
import { generatePath, useNavigate } from "react-router-dom";
import { ROUTES } from "../../utils/routes";
import { useVehicles } from "../../hooks/useVehicles";

const VehicleCard = ({ vehicle }: { vehicle: Vehicle }) => {
  const navigate = useNavigate();
  const { toggleFavourite } = useVehicles();

  const date = moment(vehicle.auctionDateTime, "YYYY/MM/DD HH:mm:ss");
  const formatted = date.format("DD MMM YYYY, HH:mm");

  const timeToAuction = getRelativeTime(new Date(formatted));

  const bidFormatter = new Intl.NumberFormat("en-GB", {
    style: "currency",
    currency: "GBP",
  });

  const updateFavourite = useCallback(
    () => toggleFavourite(vehicle.id),
    [toggleFavourite, vehicle.id],
  );

  const handleSeeDetails = () => {
    navigate(generatePath(ROUTES.VEHICLE_DETAILS, { vehicleId: vehicle.id }));
  };

  return (
    <Box
      data-testid="vehicle-card"
      display="flex"
      position="relative"
      padding="16px"
      borderRadius="30px 4px"
      gap="16px"
      sx={{
        backgroundColor: defaultColors.neutralWhite,
        border: "1px solid #f2f2f2",
      }}>
      <Box display="flex" justifyContent="center" alignItems="center">
        <img
          src={emptyPhoto}
          alt="Car"
          style={{
            width: "150px",
            height: "150px",
            objectFit: "cover",
            objectPosition: "center",
            borderRadius: "12px",
            display: "block",
            opacity: 0.7,
          }}
        />
      </Box>
      <IconButton
        onClick={updateFavourite}
        sx={{
          position: "absolute",
          top: 8,
          right: 8,
          backgroundColor: "white",
          "&:hover": {
            backgroundColor: "grey.100",
          },
        }}>
        {vehicle.favourite ? (
          <StarOutlinedIcon data-testid="favourite-icon" color="warning" />
        ) : (
          <StarBorderOutlinedIcon data-testid="not-favourite-icon" />
        )}
      </IconButton>
      <Box display="flex" flexDirection="column" gap="8px" width="100%">
        <Box>
          <Typography variant="h5" color={defaultColors.primaryMain} fontSize="20px">
            {vehicle.make} - {vehicle.model}
          </Typography>
          <Typography
            variant="body1"
            color={defaultColors.neutralDarkGrey}
            sx={{ textTransform: "capitalize" }}>
            {vehicle.fuel} | {vehicle.year} |{" "}
            {new Intl.NumberFormat("pt-PT").format(vehicle.mileage) ?? 0} Miles
          </Typography>
        </Box>
        <Divider />
        <Box display="flex" flexDirection="column" gap="8px">
          <Box>
            <Typography variant="body3" color={defaultColors.neutralDarkGrey}>
              Starting at
            </Typography>
            <Typography
              color={defaultColors.primaryLight}
              variant="body1"
              fontWeight={fontWeight.medium}>
              {bidFormatter.format(vehicle.startingBid)}
            </Typography>
          </Box>
          <Box
            display="flex"
            flexWrap="wrap"
            alignItems="center"
            gap="4px"
            justifyContent="space-between">
            <Box display="flex" alignItems="center" gap="4px">
              <TimerOutlinedIcon
                sx={{ width: "16px", heigh: "16px", color: defaultColors.primaryLight }}
              />
              <Typography color={defaultColors.primaryLight} variant="body2">
                {timeToAuction}
              </Typography>
            </Box>
            <Button
              data-testid="see-details-button"
              variant="contained"
              color="primary"
              onClick={handleSeeDetails}
              sx={{
                boxShadow: "none",
                backgroundColor: defaultColors.primaryMain,
                ":hover": {
                  boxShadow: "none",
                },
              }}>
              See details
            </Button>
          </Box>
        </Box>
      </Box>
    </Box>
  );
};

export default VehicleCard;
