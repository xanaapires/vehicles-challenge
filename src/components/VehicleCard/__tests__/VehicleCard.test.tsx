import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import { generatePath, MemoryRouter } from "react-router-dom";
import VehicleCard from "../VehicleCard";
import { vehiclesData } from "../../../assets/vehiclesData";
import { defaultStateMock } from "../../../utils/hookFixtures";
import { ROUTES } from "../../../utils/routes";

const mockedNavigate = jest.fn();
const mockUseVehicles = jest.fn();
const mockToggleFavourite = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockedNavigate,
  useParams: jest.fn(),
}));

jest.mock("../../../hooks/useVehicles", () => ({
  useVehicles: () => mockUseVehicles(),
}));

const renderComponent = () => {
  return render(
    <MemoryRouter>
      <VehicleCard vehicle={{ ...vehiclesData[0], id: 1 }} />
    </MemoryRouter>,
  );
};

describe("VehicleCard", () => {
  beforeAll(() => {
    jest.useFakeTimers();
    jest.setSystemTime(new Date("2025-08-04T00:00:00Z"));
  });

  afterAll(() => {
    jest.useRealTimers();
  });

  test("Renders VehicleCard correctly", () => {
    mockUseVehicles.mockReturnValue({ ...defaultStateMock });

    const { container } = renderComponent();

    expect(container).toMatchSnapshot();
    expect(screen.getByTestId("vehicle-card")).toBeVisible();
  });

  test("Handles See Details correctly", () => {
    mockUseVehicles.mockReturnValue({ ...defaultStateMock });

    renderComponent();

    expect(screen.getByTestId("vehicle-card")).toBeVisible();
    expect(screen.getByTestId("see-details-button")).toBeVisible();

    fireEvent.click(screen.getByTestId("see-details-button"));

    expect(mockedNavigate).toHaveBeenCalledWith(
      generatePath(ROUTES.VEHICLE_DETAILS, { vehicleId: 1 }),
    );
  });

  test("Handles toogle favourite correctly", () => {
    mockUseVehicles.mockReturnValue({
      ...defaultStateMock,
      toggleFavourite: mockToggleFavourite,
    });

    renderComponent();

    expect(screen.getByTestId("vehicle-card")).toBeVisible();
    expect(screen.getByTestId("favourite-icon")).toBeVisible();

    fireEvent.click(screen.getByTestId("favourite-icon"));
    expect(mockToggleFavourite).toHaveBeenCalled();
  });
});
