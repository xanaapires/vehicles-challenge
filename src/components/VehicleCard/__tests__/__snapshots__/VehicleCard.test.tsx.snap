// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VehicleCard Renders VehicleCard correctly 1`] = `
<div>
  <div
    class="MuiBox-root css-10trl62"
    data-testid="vehicle-card"
  >
    <div
      class="MuiBox-root css-1n2mv2k"
    >
      <img
        alt="Car"
        src="empty_photo.jpg"
        style="width: 150px; height: 150px; object-fit: cover; object-position: center; border-radius: 12px; display: block; opacity: 0.7;"
      />
    </div>
    <button
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1u2l4gi-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-colorWarning MuiSvgIcon-fontSizeMedium css-3adcp6-MuiSvgIcon-root"
        data-testid="favourite-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <div
      class="MuiBox-root css-1hqz9cp"
    >
      <div
        class="MuiBox-root css-0"
      >
        <h5
          class="MuiTypography-root MuiTypography-h5 css-o8v361-MuiTypography-root"
        >
          Toyota
           - 
          C-HR
        </h5>
        <p
          class="MuiTypography-root MuiTypography-body1 css-3uxmf5-MuiTypography-root"
        >
          diesel
           | 
          2022
           |
           
          743
           Miles
        </p>
      </div>
      <hr
        class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
      />
      <div
        class="MuiBox-root css-1821gv5"
      >
        <div
          class="MuiBox-root css-0"
        >
          <span
            class="MuiTypography-root MuiTypography-body3 css-jouwp7-MuiTypography-root"
          >
            Starting at
          </span>
          <p
            class="MuiTypography-root MuiTypography-body1 css-1ap16v6-MuiTypography-root"
          >
            Â£17,000.00
          </p>
        </div>
        <div
          class="MuiBox-root css-1ffsm0h"
        >
          <div
            class="MuiBox-root css-171onha"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-2d4n43-MuiSvgIcon-root"
              data-testid="TimerOutlinedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15 1H9v2h6zm-4 13h2V8h-2zm8.03-6.61 1.42-1.42c-.43-.51-.9-.99-1.41-1.41l-1.42 1.42C16.07 4.74 14.12 4 12 4c-4.97 0-9 4.03-9 9s4.02 9 9 9 9-4.03 9-9c0-2.12-.74-4.07-1.97-5.61M12 20c-3.87 0-7-3.13-7-7s3.13-7 7-7 7 3.13 7 7-3.13 7-7 7"
              />
            </svg>
            <p
              class="MuiTypography-root MuiTypography-body2 css-1cwj2yy-MuiTypography-root"
            >
              in 133 day(s) and 9 hour(s)
            </p>
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-9wvbqu-MuiButtonBase-root-MuiButton-root"
            data-testid="see-details-button"
            tabindex="0"
            type="button"
          >
            See details
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
